app/services/fees/create_from_usage_threshold_service.rb

```diff
@@ -0,0 +1,59 @@
+# frozen_string_literal: true
+
+module Fees
+  class CreateFromUsageThresholdService < BaseService
+    def initialize(usage_threshold:, invoice:, amount_cents:)
+      @usage_threshold = usage_threshold
+      @invoice = invoice
+      @amount_cents = amount_cents
+
+      super
+    end
+
+    def call
+      fee = Fee.new(
+        subscription: invoice.subscriptions.first,
+        invoice:,
+        usage_threshold:,
+        invoice_display_name: usage_threshold.threshold_display_name,
+        invoiceable: usage_threshold,
+        amount_cents: amount_cents,
+        amount_currency: invoice.currency,
+        fee_type: :progressive_billing,
+        units:,
+        unit_amount_cents: unit_amount_cents,
+        payment_status: :pending,
+        taxes_amount_cents: 0,
+        properties: {
+          charges_from_datetime: invoice.invoice_subscriptions.first.charges_from_datetime,
+          charges_to_datetime: invoice.invoice_subscriptions.first.charges_to_datetime,
+          timestamp: invoice.invoice_subscriptions.first.timestamp
+        }
+      )
+
+      taxes_result = Fees::ApplyTaxesService.call(fee:)
+      taxes_result.raise_if_error!
+
+      fee.save!
+      result.fee = fee
+
+      result
+    end
+
+    private
+
+    attr_reader :usage_threshold, :invoice, :amount_cents
+
+    def units
+      return 1 unless usage_threshold.recurring?
+
+      amount_cents.fdiv(usage_threshold.amount_cents)
```

Comment:
more a question than a remark, but this means we can have for example 0.8 units? (I think that's fine, just wondering if this will be confusing :smile)

Reply:
Theoretically yes, but actually the values:
```ruby
units = (total_lifetime_usage_amount_cents - invoiced_threshold_amount_cents) / usage_threshold.amount_cents
units * usage_threshold.amount_cents
``` 
so it should always be an integer